cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedInitialization.qll

```diff
@@ -115,10 +115,14 @@ abstract class TranslatedVariableInitialization extends TranslatedElement, Initi
    * evaluating the initializer.
    */
   final predicate hasUninitializedInstruction() {
-    not exists(getInitialization()) or
-    getInitialization() instanceof TranslatedListInitialization or
-    getInitialization() instanceof TranslatedConstructorInitialization or
-    getInitialization().(TranslatedStringLiteralInitialization).zeroInitRange(_, _)
+    (
+      not exists(getInitialization()) or
+      getInitialization() instanceof TranslatedListInitialization or
+      getInitialization() instanceof TranslatedConstructorInitialization or
+      getInitialization().(TranslatedStringLiteralInitialization).zeroInitRange(_, _)
+    ) and
+    // Variables with static or thread-local storage duration are zero-initialized at program startup.
```

Comment:
Should there be an explicit zero-initialization for dynamically initialized static variables with a nontrivial constructor?

Reply:
I guess I'm assuming that any analysis that cares enough about that would also already be handling the fact that all variables with static storage duration are zero-initialized. I think that having the dynamic initialization in the IR is more about ensure that we _don't_ appear to _always_ initialize the variable on every execution, rather than actually giving good information about the possible values of the variable. Anybody else have an opinion here?

Reply:
> giving good information about the possible values of the variable

I think that's quite a central task for the IR. But until it can also do that for ordinary globals, I don't see why it should be done for static locals.